File: solutions/haskell/perfect-numbers/1/src/PerfectNumbers.hs
module PerfectNumbers (classify, Classification (..)) where

data Classification = Deficient | Perfect | Abundant deriving (Eq, Show)

classify :: Int -> Maybe Classification
classify n =
  if n <= 0
    then Nothing
    else
      let ss = s n
       in if ss == n
            then Just Perfect
            else
              if ss < n
                then Just Deficient
                else Just Abundant

s :: Int -> Int
s n = sum $ filter (\x -> mod n x == 0) [1 .. (n - 1)]

File: solutions/csharp/perfect-numbers/1/PerfectNumbers.cs
public enum Classification
{
    Perfect,
    Abundant,
    Deficient
}

public static class PerfectNumbers
{
    public static Classification Classify(int number)
    {
        if (number <= 0)
        {
            throw (new ArgumentOutOfRangeException());
        }
        else if (number == 1)
        {
            return Classification.Deficient;
        }
        var i = 1;
        var ans = 0;
        while (i * i < number)
        {
            if (number % i == 0)
            {
                ans += i + number / i;
            }
            i += 1;
        }
        if (i * i == number)
        {
            ans += i;
        }
        ans -= number;
        if (ans == number)
        {
            return Classification.Perfect;
        }
        else if (ans < number)
        {
            return Classification.Deficient;
        }
        else
        {
            return Classification.Abundant;
        }
    }
}
